### ISSUE TITLE
systematic pattern in the residuals between observed and predicted spot Z positions

### ISSUE THREAD

**Issue by dagewa** [dials/dials#591]
First noted in #457, but not really part of that issue, so I made a new issue here. Taking just the first few images of the [thaumatin BAG training data](https://zenodo.org/record/10271#.WzyjgHYzrCI)
```
dials.import th_8_2_000{1..9}.cbf
dials.find_spots datablock.json
dials.index datablock.json strong.pickle
dials.report experiments.json indexed.pickle
```
Produces a plot like this with a weird "sawtooth" shape
![newplot](https://user-images.githubusercontent.com/11502083/42286625-6f35dff2-7fab-11e8-9720-7c6aaf2fc79e.png)

I thought I understood this and that it was due to reflections that are present on only one image. However, we can remove those reflections completely:
```
from dials.array_family import flex
rt=flex.reflection_table.from_pickle('indexed.pickle')

# remove reflections with a single image shoebox
shoes=rt['shoebox']
sel=flex.bool([s.zsize() > 1 for s in shoes])
rt = rt.select(sel)

# save filtered reflections
rt.as_pickle('indexed_filtered.pickle')
```
Now index again to ensure refinement against the reduced set of reflections
```
dials.index datablock.json indexed_filtered.pickle
dials.report experiments.json indexed.pickle
```
There is still some kind of sawtooth pattern, though not quite as clear (but there are fewer reflections too)
![newplot 1](https://user-images.githubusercontent.com/11502083/42286833-fb0638ba-7fab-11e8-9fde-531fcb6f15a3.png)
I don't understand this. Has anybody got any ideas?

**Comment by dagewa** [comment]
@biochem-fan suggested recycling the centroids after integration to see what effect that has. Starting with the original `indexed.pickle` (not the one with 1 image reflections removed as above):
```
dials.refine experiments.json indexed.pickle
dials.refine refined_experiments.json refined.pickle
dials.integrate refined_experiments.json refined.pickle
dev.dials.filter_reflections integrated.pickle inclusions.flag=used_in_refinement
dials.refine filtered.pickle integrated_experiments.json
dials.report refined.pickle refined_experiments.json
```
The `dev.dials.filter_reflections` step ensures we only use reflections originally used in refinement, otherwise very weak reflections are included in the second round and RMSDs look much worse. After this second round of refinement. the sawtooth effect is still visible.
![newplot 2](https://user-images.githubusercontent.com/11502083/42313632-1ea61ddc-803b-11e8-839c-1fefc87a9864.png)

**Comment by graeme-winter** [comment]
I think this is principally a representation of data issue....

```
from dials.array_family import flex
import matplotlib
matplotlib.use('Agg')

from matplotlib import pyplot


refl = flex.reflection_table.from_pickle('indexed.pickle')
z_obs = refl['xyzobs.px.value'].parts()[2]
z_cal = refl['xyzcal.px'].parts()[2]
dz_all = z_obs - z_cal

fig, ax1 = pyplot.subplots()
ax1.set_xlim(-2.5, 2.5)
ax1.set_xlabel('$\\delta Z$')
ax1.set_ylabel('$N$')

ax1.hist(dz_all.as_numpy_array(), bins=200, range=(-2.5, 2.5), label='All')

sel = flex.bool([s.zsize() > 1 for s in refl['shoebox']])
dz_all = dz_all.select(sel)

ax1.hist(dz_all.as_numpy_array(), bins=200, range=(-2.5, 2.5), label='> 1')

handles, labels = ax1.get_legend_handles_labels()
pyplot.legend(handles=handles, loc=1)

fig.tight_layout()
pyplot.savefig('dz_histogram.pdf')
```

**Comment by graeme-winter** [comment]
[dz_histogram.pdf](https://github.com/dials/dials/files/2178823/dz_histogram.pdf)

Almost all of the deltas are +/- a very small amount => the images above are misleading

**Comment by graeme-winter** [comment]
![dz_histogram](https://user-images.githubusercontent.com/1241716/42489585-baf6f744-8403-11e8-9f2f-f00693e9cc11.png)

Histogram in helpful

**Comment by graeme-winter** [comment]
<img width="1243" alt="screen shot 2018-07-10 at 07 13 53" src="https://user-images.githubusercontent.com/1241716/42492397-dba47b8a-8410-11e8-80df-25211cfad6c3.png">

All duff reflections around rotation axis => no mystery here

**Comment by graeme-winter** [comment]
@dagewa I suspect you are experiencing the power of the human eye for pattern recognition - this is a very weak signal (in truth) instead. Only real thing which occurs to me is that this is really a discretisation error ultimately...

**Comment by ndevenish** [comment]
> but all the values in the histogram above are > 0

As discussed in hangouts, this seems to be the translation to plotly bug fixed by @benjaminhwilliams in 0109af821efce5f630234e518f152be71449c9ca?

**Comment by graeme-winter** [comment]
![newplot 6](https://user-images.githubusercontent.com/1241716/42500782-10c0eb0a-842a-11e8-95d4-ff8724a7f61f.png)

Updated plot - sensible now

**Comment by graeme-winter** [comment]
Fascinating discussion with @ndevenish looks exactly like the discretisation error as expected - Nick's code:

```
#!/usr/bin/env python

import matplotlib.pyplot as plt
import numpy as np
import cPickle as pickle
import enum
import dials.array_family.flex as flex
import sys

if "-h" in sys.argv or "--help" in sys.argv or len(sys.argv[1:]) != 1:
    print("Usage: plot_centroid_per_shoebox.py <indexed.pickle>")
    sys.exit(0)

with open(sys.argv[1]) as f:
    reflections = pickle.load(f)

used_in_refine = reflections.get_flags(reflections.flags.used_in_refinement)
refined = reflections.select(used_in_refine)
# not_refined = reflections.select(~used_in_refine)

plt.figure()
plt.hist([x.zsize() for x in reflections["shoebox"]], bins=8, range=(1,9))
plt.xlabel("Shoebox width")
plt.ylabel("N")

plt.figure()
plt.suptitle("Centroid differences for reflections that span N images")
for n in range(1,7):
    plt.subplot(2,3,n)
    select = flex.bool([x.zsize()==n for x in refined["shoebox"]])
    refs = refined.select(select)

    _, _, zc = [x.as_numpy_array() for x in refs["xyzcal.px"].parts()]
    _, _, zo = [x.as_numpy_array() for x in refs["xyzobs.px.value"].parts()]
    zd = zo - zc
    plt.hist2d(zc%1, zd, range=[[0,1],[-1,1]], bins=[30,30])
    plt.xlim(0, 1)
    plt.ylim(-0.5, 0.5)
    plt.title(n)
    plt.xlabel("$Z_o \% 1$")
    plt.ylabel("$Z_o-Z_c$")
plt.tight_layout()

plt.show()
```

shows the same phenomenon with other data sets

**Comment by graeme-winter** [comment]
![figure_2](https://user-images.githubusercontent.com/1241716/42507941-ca0040b6-843e-11e8-94f0-7d4ffbda174f.png)

**Comment by graeme-winter** [comment]
Odd sized shoeboxes have central tendencies, even sized biased towards extrema as may be expected => see also "tails" correction

**Comment by graeme-winter** [comment]
This above is CDK2 sample 6 4% data set

**Comment by ndevenish** [comment]
For IO4 Thaumatin training set:
![discrete_error](https://user-images.githubusercontent.com/529963/42508121-4e926dae-843f-11e8-9390-8b1c1291ae51.png)

**Comment by dagewa** [comment]
Thanks all for looking at this. Could somebody please write a summary here to aid clarity and explain what these plots show? Cheers, D

**Comment by graeme-winter** [comment]
Yeh, will do - was thinking in terms of a short phenix newsletter write up or something - exploring phenomena with DIALS - but would be good to include a conclusion here as well. 

Commentary at the moment goes like:

6 plots - data are separated into reflections with shoebox depth 1…6 and then the Z centroid offset plotted vs. Z modulo 1 on the histogram - shows clear pattern for depth = 1, 2, 3 of the sawtooth pattern you identified, and suggests (to this viewer at least) that this is a discretisation error.

**Comment by ndevenish** [comment]
Just noticed an axis labelling error: X-axis is Z_calculated, not observed. Otherwise the 1-shoebox graph makes no sense (all would be integer).

**Comment by biochem-fan** [comment]
Hi Graeme, wonderful analysis!

This happens because tails of the rocking curve are weak and not flagged as strong pixels, so centroids are biased towards frames that contain the centre of the curve. Is my understanding correct?

**Comment by graeme-winter** [comment]
By and large, yes - I think this captures the same basic issue as the tails correction in scala

It was fun to look into, and makes a nice demonstration of the value of a toolkit for data processing rather than a black box 🙂

**Comment by dagewa** [comment]
Thanks all for looking into this. I suggest we keep this issue open for a little longer while we are still looking into it, with the intention of presenting the issue, in a newsletter or slides. There may be a few points still to explore, for example I know @ndevenish was still exploring the effect or recycling centroids after integration.
